core/src/server/subscription.rs

```diff
@@ -0,0 +1,475 @@
+pub type Subscribers = Arc<Mutex<FxHashMap<SubscriptionKey, (MethodSink, mpsc::Receiver<()>)>>>;
+
+/// Convert something into a subscription close notification
+/// before a subscription is terminated.
+pub trait IntoSubscriptionResponse {
+	/// Convert something into a subscription response
+	fn into_response(self) -> SubscriptionCloseResponse;
+}
+
+/// Represents what action that will sent when a subscription callback returns.
+#[derive(Debug)]
+pub enum SubscriptionCloseResponse {
+	/// No further message will be sent.
+	None,
+	/// Send a ordinary subscription response.
+	Some(SubscriptionMessage),
+	/// Send a subscription error response.
+	Err(SubscriptionMessage),
+}
+
+impl<T> IntoSubscriptionResponse for Option<T>
+where
+	T: serde::Serialize,
+{
+	fn into_response(self) -> SubscriptionCloseResponse {
+		match self {
+			Some(msg) => match SubscriptionMessage::from_json(&msg) {
+				Ok(m) => SubscriptionCloseResponse::Some(m),
+				Err(e) => SubscriptionCloseResponse::Err(e.to_string().into()),
+			},
+			None => SubscriptionCloseResponse::None,
+		}
+	}
+}
+
+impl<T, E> IntoSubscriptionResponse for Result<T, E>
+where
+	T: serde::Serialize,
+	E: ToString,
+{
+	fn into_response(self) -> SubscriptionCloseResponse {
+		match self {
+			Ok(msg) => match SubscriptionMessage::from_json(&msg) {
+				Ok(m) => SubscriptionCloseResponse::Some(m),
+				Err(e) => SubscriptionCloseResponse::Err(e.to_string().into()),
+			},
+			Err(e) => SubscriptionCloseResponse::Err(e.to_string().into()),
+		}
+	}
+}
+
+/// A complete subscription message or partial subscription message.
+#[derive(Debug, Clone)]
+pub enum SubscriptionMessageInner {
+	/// Complete JSON message.
+	Complete(String),
+	/// Need subscription ID and method name.
+	NeedsData(String),
+}
+
+/// Subscription message.
+#[derive(Debug, Clone)]
+pub struct SubscriptionMessage(pub(crate) SubscriptionMessageInner);
+
+impl SubscriptionMessage {
+	/// Create a new subscription message from JSON.
+	///
+	/// Fails if the value couldn't be serialized.
+	pub fn from_json(t: &impl Serialize) -> Result<Self, serde_json::Error> {
+		serde_json::to_string(t).map(|json| SubscriptionMessage(SubscriptionMessageInner::NeedsData(json)))
+	}
+
+	pub(crate) fn from_complete_message(msg: String) -> Self {
+		SubscriptionMessage(SubscriptionMessageInner::Complete(msg))
+	}
+
+	pub(crate) fn empty() -> Self {
+		Self::from_complete_message(String::new())
+	}
+}
+
+impl<T> From<T> for SubscriptionMessage
+where
+	T: AsRef<str>,
+{
+	fn from(msg: T) -> Self {
+		SubscriptionMessage(SubscriptionMessageInner::NeedsData(format!("\"{}\"", msg.as_ref())))
+	}
+}
+
+/// Represent a unique subscription entry based on [`SubscriptionId`] and [`ConnectionId`].
+#[derive(Clone, Debug, PartialEq, Eq, Hash)]
+pub struct SubscriptionKey {
+	pub(crate) conn_id: ConnectionId,
+	pub(crate) sub_id: SubscriptionId<'static>,
+}
+
+#[derive(Debug, Clone, Copy)]
+pub(crate) enum SubNotifResultOrError {
+	Result,
+	Error,
+}
```

Comment:
Nit: since `Result` is a thing with an `Ok` or `Err` variant already, maybe `Response, Error` makes more sense here than `Result, Error` for the variant names?